### PR TITLE
Feature - use behaviours

### PR DESCRIPTION
**Pull request by joefitter**
This PR removes the dependency on hof-controllers and instead installs hof-form-controller ^3.0.1 which already has hooks and session IO mixed in. Also installed mixwith.js so steps can now provide one or more custom "behaviours" in place of a singular custom controller. Custom controllers are still supported but a deprication warning is now shown if a custom controller is used.

### CODE REVIEW

**Review by easternbloc — APPROVED** [review]
LGTM

**Comment by joefitter** [comment]
@lennym would like your input if you get time, tomorrow is fine though

File: README.md [hunk 1]
```diff
@@ -65,7 +65,7 @@ The minimum amount of configuration for a wizard step is the `next` property to
 * `fields` - specifies which of the fields from the field definition list are applied to this step. Form inputs which are not named on this list will not be processed. Default: `[]`
 * `template` - Specifies the template to render for GET requests to this step. Defaults to the route (without trailing slash)
 * `backLink` - Specifies the location of the step previous to this one. If not specified then an algorithm is applied which checks the previously visited steps which have the current step set as `next`.
-* `controller` - The constructor for the controller to be used for this step's request handling. The default is an extension of the [hof-form-controller](https://www.npmjs.com/package/hof-form-controller), which is exported as a `Controller` property of this module. If custom behaviour is required for a particular form step then custom extensions can be defined - see [Custom Controllers](#custom-controllers)
+* `behaviours` - A single behaviour, or an array of behaviours to be mixed into the base controller to extend functionality.
```
Comment (lennym) [inline README.md]: Is the order of behaviours significant when passed as an array? My guess is yes, so I'd probably want to know if they are going to be applied left-to-right or right-to-left.

File: README.md [hunk 2]
```diff
@@ -80,34 +80,37 @@ A number of options can be passed to the wizard as a third argument to customise
...
-
-module.exports = CustomController
+module.exports = SuperClass => class extends SuperClass {
+  constructor() {
+    super();
+    this.use((req, res, next) => {
+      console.log(req.method, req.url);
+      next();
+    });
+  }
```
Comment (lennym) [inline README.md]: I don't feel like this is necessarily the *best* example as it extends the constructor, which isn't a common case. I'd either change it to one with a method extension or at least add an example with method extension.

File: lib/wizard.js [hunk 3]
```diff
@@ -2,10 +2,22 @@
...
 
+const createController = (SuperClass, behaviours) => {
+  /*
+   * This class declaration could be better written using
+   * array spread syntax, supported in node >= 5.11.0:
+   *
+   * class Controller extends mix(SuperClass).with(...behaviours) {}
+   */
+  let _mix;
+  return class extends (_mix = mix(SuperClass)).with.apply(_mix, behaviours) {};
```
Comment (lennym) [inline lib/wizard.js]: This feels a little like "one line for the sake of being one line". Looking at it, I think this is equivalent and a bit clearer:

```js
const _mix = mix(SuperClass);
return class extends _mix.with.apply(_mix, behaviours) {};
```

**Review by lennym — CHANGES_REQUESTED** [review]

**Comment by lennym** [comment]
Looks pretty good overall. I just have a few minor points and clarifications. The biggest thing for me is being clear on what order behaviours get mounted in.

File: package.json [hunk 4]
```diff
@@ -28,17 +28,19 @@
     "csrf": "^3.0.2",
     "debug": "^2.1.2",
     "express": "^4.12.2",
+    "hof-form-controller": "^3.0.1",
     "hof-model": "^2.0.0",
-    "hof-controllers": "^6.0.0",
```
Comment (lennym) [inline package.json]: Does this now make `hof-controllers` officially legacy?

Comment (lennym) [reply package.json]: :tada:

File: README.md (hunk 2 above)
Comment (joefitter) [reply README.md]: Yeah I agree - I was being lazy and just modified the existing example - I'll change it

File: lib/wizard.js (hunk 3 above)
Comment (joefitter) [reply lib/wizard.js]: whatswrongwithcrammingeverythingtogether?

**Comment by joefitter** [comment]
@lennym howzat?

File: package.json [hunk 5]
```diff
@@ -28,17 +28,19 @@
...
-    "lodash": "^4.17.2"
+    "lodash": "^4.17.2",
+    "mixwith": "^0.1.1"
   },
   "devDependencies": {
     "chai": "^3.5.0",
     "eslint-config-homeoffice": "^2.0.0",
     "istanbul": "^0.4.3",
     "mocha": "^3.1.2",
+    "proxyquire": "^1.7.11",
```
Comment (lennym) [inline package.json]: This doesn't seem to be used anywhere - or at least hasn't been added elsewhere in this PR - is it a linting thing?

**Review by lennym — CHANGES_REQUESTED** [review]

File: package.json (hunk 5 above)
Comment (easternbloc) [reply package.json]: You're right this shouldn't be here it's not used.

Comment (joefitter) [reply package.json]: nope it aint no more 🔥

File: README.md [hunk 6]
```diff
@@ -65,7 +65,7 @@ The minimum amount of configuration for a wizard step is the `next` property to
...
 * `template` - Specifies the template to render for GET requests to this step. Defaults to the route (without trailing slash)
 * `backLink` - Specifies the location of the step previous to this one. If not specified then an algorithm is applied which checks the previously visited steps which have the current step set as `next`.
-* `controller` - The constructor for the controller to be used for this step's request handling. The default is an extension of the [hof-form-controller](https://www.npmjs.com/package/hof-form-controller), which is exported as a `Controller` property of this module. If custom behaviour is required for a particular form step then custom extensions can be defined - see [Custom Controllers](#custom-controllers)
+* `behaviours` - A single behaviour, or an array of behaviours to be mixed into the base controller to extend functionality. If an array of behaviours is given, they are applied left-to-right, so calling super in the right-most behaviour will point to the previous behaviour in the array.
```
Comment (lennym) [inline README.md:68]: If we're going to deprecate `controller` as a step option then maybe it's a good idea to use npmjs.com/deprecate to output a message if one is passed.

Comment (easternbloc) [reply README.md:68]: Seems like a nice idea.

Comment (joefitter) [reply README.md:68]: ah yeah, that's nicer - I just added a couple of `console.warns`

Comment (lennym) [reply README.md:68]: Oh right, I hadn't spotted that. `console.warn` is *probably* fine, `deprecate` just is a bit less spammy and a bit more colourful.

**Comment by joefitter** [comment]
@lennym @JoeChapman @easternbloc updated - I removed the test as its _just_ for a deprecation warning - stubbing the console was trivial, but stubbing deprecate would require re-installing proxyquire, which seems like overkill for a warning

**Comment by lennym** [comment]
That is 100% fine by me.

**Review by lennym — APPROVED** [review]

**Comment by easternbloc** [comment]
@joefitter i love that guy he's the happiest little birthday person on the planet

**Comment by joefitter** [comment]
It is literally my favourite gif ever

**Pull request merged**